File: Day_12/Day12.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <map>
#include <set>
using namespace std;

string stringToLower(string s)
{
    string lowerS = "";
    for (char ch : s)
        lowerS += tolower(ch);
    return lowerS;
}

int countPaths(map<string, vector<string>> &adjList, string src, vector<string> path, set<string> visitedSmallCaves)
{
    if (src == "end")
        return 1;

    if (stringToLower(src) == src //
        && visitedSmallCaves.find(src) != visitedSmallCaves.end())
        return 0;

    if (stringToLower(src) == src)
        visitedSmallCaves.insert(src);

    int count = 0;
    for (string dst : adjList[src])
    {
        path.push_back(dst);
        count += countPaths(adjList, dst, path, visitedSmallCaves);
        path.pop_back();
    }
    return count;
}

int countPathsModified(map<string, vector<string>> &adjList, string src, vector<string> path,
                       set<string> visitedSmallCaves, bool visitedSmallCaveTwice)
{
    if (stringToLower(src) == src                                 //
        && visitedSmallCaves.find(src) != visitedSmallCaves.end() //
        && (src == "start" || visitedSmallCaveTwice))
        return 0;

    if (src == "end")
        return 1;

    if (src != "start" && stringToLower(src) == src //
        && visitedSmallCaves.find(src) != visitedSmallCaves.end())
        visitedSmallCaveTwice = true;

    if (stringToLower(src) == src)
        visitedSmallCaves.insert(src);

    int count = 0;
    for (string dst : adjList[src])
    {
        path.push_back(dst);
        count += countPathsModified(adjList, dst, path, visitedSmallCaves, visitedSmallCaveTwice);
        path.pop_back();
    }
    return count;
}

map<string, vector<string>> getInput()
{
    map<string, vector<string>> adjMap;
    ifstream inputFile("./Day12_Input.txt");
    string line;
    while (inputFile >> line)
    {
        int pos = line.find("-");
        string src = line.substr(0, pos),
               dst = line.substr(pos + 1);

        adjMap[src].push_back(dst);
        adjMap[dst].push_back(src);
    }
    return adjMap;
}

int main()
{
    map<string, vector<string>> adjMap;
    adjMap = getInput();

    // Part 1
    int count = countPaths(adjMap, "start", vector<string>{"start"}, set<string>());
    cout << "Number of Paths from start -> end = " << count << endl;

    // Part 2
    count = countPathsModified(adjMap, "start", vector<string>{"start"}, set<string>(), false);
    cout << "Number of Paths from start -> end visiting any 1 small cave twice = " << count << endl;

    return 0;
}

File: Day_12/Day12.py
def getPaths(adjList, path, src):
    if src.islower() and src in path[:-1]:
        return 0

    if src == 'end':
        return 1

    count = 0
    for dst in adjList[src]:
        path.append(dst)
        count += getPaths(adjList, path, dst)
        path.pop()
    return count


def getModifiedPaths(adjList, path, src, smallTwiceVisited):
    if src.islower() and smallTwiceVisited and src in path[:-1]:
        return 0

    if src == 'end':
        return 1

    if src.islower() and src in path[:-1]:
        smallTwiceVisited = True

    count = 0
    for dst in adjList[src]:
        path.append(dst)
        if dst != 'start':
            count += getModifiedPaths(adjList, path, dst, smallTwiceVisited)
        path.pop()
    return count


def getInput():
    adjList = {}
    with open('./Day12_Input.txt') as InputFile:
        for line in InputFile:
            if '\n' in line:
                line = line.strip()
            token = line.split('-')

            if token[0] in adjList.keys():
                adjList[token[0]].append(token[1])
            else:
                adjList[token[0]] = [token[1]]

            if token[1] in adjList.keys():
                adjList[token[1]].append(token[0])
            else:
                adjList[token[1]] = [token[0]]
    return adjList


adjList = getInput()

# part 1
countPath = getPaths(adjList, ['start'], 'start')
print('Number of Paths from start -> end = ' + str(countPath))

# part 2
countPath = getModifiedPaths(adjList, ['start'], 'start', False)
print('Number of Paths from start -> end visiting any 1 small cave twice = ' + str(countPath))
